### DISCUSSION TITLE
How to get the variable values from neighbor elements of current element？

### DISCUSSION THREAD

**Discussion by xiaojbing** [idaholab/moose#21907] [Q&A General] [answered]
Hello，
   I want to know how to get the  variable from  nieghbor element, and which class need to be inherited. 
thank you.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Could you please give us more details about what you're trying to do here?
Typically you would want to use the coupledNeighborValue from the Coupleable interface to retrieve the neighbor values.

Guillaume

**Reply by xiaojbing** [reply]
Thank you.

I want to caculate and refresh a coefficient every timestep.
This coefficient rely an average value of current element and surrounding element. 
For example,  a is a value stored in current element , b , c , d are sepretedly the values stored in surrounding elements of current element. and I want to get 1/4(a+b+c+d).
What should I do?

**Reply by GiudGiud** [reply]
so is there a single value by element? Many variable types will have different values at each quadrature point

**Reply by xiaojbing** [reply]
Thank you

the a b c d are the same variable, just in different element. An element only have one value, for example, **A** represent velocity, each element only have one velocity. I want to get an average velocity of current element and it's surrouding element. may not the value 
at quadrature point. So how to get the single value in different element?

**Reply by lindsayad** [reply]
For a generic finite element formulation there is no one single variable value in an element. Only constant monomial finite element types have constant elemental values

**Reply by xiaojbing** [reply]
I am a new user of moose....

for example,
In ex_01 , every quadrature point have a variable `diffused`，the value at other positions is interpolated, right?
```
[Variables]
  [./diffused]
    order = FIRST
    family = LAGRANGE
  [../]
[]
```

**Reply by GiudGiud** [reply]
yes.
You may project the variable on a CONSTANT MONOMIAL basis if you want need a cell-average

**Reply by xiaojbing** [reply]
Thank you.
Another question is how to get the variable in elements around the current element?
Can I use a pointer to get the sourrounding elemental information? pointer ` _current_elem` in Kernel point to current element, right?

**Reply by GiudGiud** [reply]
From current elem you can obtain the list of neighbors. From there you could obtain the degree-of-freedom (dof) index and index into the solution variable. I dont think we have any open example of doing this though

**Reply by GiudGiud** [reply]
it's pretty hacky at this point

**Reply by xiaojbing** [reply]
Thank you.
That is good, but not convient to get the variable value in cornel element.
![image](https://user-images.githubusercontent.com/87591264/186558753-d4e9b644-1fdf-452d-9ece-a0b274ea7def.png)

So I want to ask a new question...
The variable values at each element node are interplated. which quadrature points are used to caculate the value at node? The element inner quadrature points around the node? if so, I want to to get the variable value at every current element node. And my question will be solved.  
![image](https://user-images.githubusercontent.com/87591264/186563868-2ffa1a51-613f-48fe-9a2b-c7ff57974c6e.png)

**Reply by GiudGiud** [reply]
the projection is done automatically if you use nodal variables everywhere, since nodal variable QPs are the nodes themselves

**Reply by xiaojbing** [reply]
So, how to get the nodal variables value?

**Reply by GiudGiud** [reply]
same way. Access through the dofs.
From _current_node you can get the local dofs.

For the neighbors though I m not sure. I dont see anything in the Node doxygen to do that.
